lib/model/store.dart

```diff
@@ -472,4 +475,21 @@ class LivePerAccountStore extends PerAccountStore {
       }
     }
   }
+
+  /// Send this client's notification token to the server, now and if it changes.
+  ///
+  /// TODO(#321) handle iOS/APNs; currently only Android/FCM
+  // TODO(#322) save acked token, to dedupe updating it on the server
+  // TODO(#323) track the registerFcmToken/etc request, warn if not succeeding
+  Future<void> registerNotificationToken() async {
+    // TODO call removeListener on [dispose]
+    NotificationService.instance.token.addListener(_registerNotificationToken);
+    await _registerNotificationToken();
+  }
+
+  Future<void> _registerNotificationToken() async {
+    final token = NotificationService.instance.token.value;
+    if (token == null) return;
```

Comment:
Ah interesting; I guess this early return doesn't make things fall down if the get-device-token request happens to complete after the `await _registerNotificationToken()` (above) is called. That's because of this listener we add:

```dart
NotificationService.instance.token.addListener(_registerNotificationToken);
```

With that, it looks like we'll be sure to dispatch the register-with-server request as soon as we have a good result from the get-device-token request, which is good.

(And that behavior is covered in the tests; nice!)

Still, although that race is handled for users' benefit, I wonder if `_registerNotificationToken` and its public wrapper `registerNotificationToken` have an unnecessarily misleading return type. I think their callers can't tell by awaiting the Future whether the token registration succeeded, failed (as with an API error), or is still pending. Is that right? Maybe `void` as a return type is a better match for that behavior.

Reply:
> Maybe `void` is the right return type given that.

I see that the tests use the `Future<void>` return type by `await`ing the returned Future. Instead of doing that, I wonder if e.g. [`FakeAsync.elapse`](https://pub.dev/documentation/fake_async/latest/fake_async/FakeAsync/elapse.html) could be used instead?

Reply:
Yeah, this return type actually was `void` in a draft before I wrote the tests.



> Instead of doing that, I wonder if e.g. [`FakeAsync.elapse`](https://pub.dev/documentation/fake_async/latest/fake_async/FakeAsync/elapse.html) could be used instead?

It's a bit tricky because in the "initially unknown" test case, we first want to wait for `store.registerNotificationToken` to complete whatever it's going to do, then check no request was made, and only after that wait for `NotificationService.start` to finish, including its `getToken` call.

I think a fully robust solution probably involves extending TestZulipBinding so that the test can control when `getToken` finishes.